### PR TITLE
Add listing student groups feature in DG

### PR DESCRIPTION
**Pull request by teekoksiang**

### CODE REVIEW

File: docs/DeveloperGuide.md [hunk 1]
```diff
@@ -235,6 +235,33 @@ Pros | Cons
...
+#### 3.2.1 Implementation
+The listing all student groups feature is facilitated by `ListGroupCommand`. It extends `Command` and overrides `Command#execute()` to list all student groups.
+
+Given below is a step-by-step usage scenario and how the listing all student groups feature works:
+
+1. The user launches the application and types `list-group` into the input box.
+2. The `UI` handles the input and calls `LogicManager#execute()` to execute it.
+3. The `AddressBookParser#parseCommand()` is called to parse the input and return a `ListGroupCommand`.
+4. The `ListGroupCommand` calls `ListGroupCommand#execute()` to retrieve the list of all residents by calling `Model#getFilteredPersonList()`.
+5. The `ListGroupCommand#execute()` collects all residents' student groups and return the results to `UI`.
```
Comment (jieffyy) [inline docs/DeveloperGuide.md]: Hmm... These 2 steps are not very clear.

My guess is that `ListGroupCommand#execute` gets the `filteredpersonlist`, then iterates through it and gets a set of student groups?

File: docs/DeveloperGuide.md [hunk 2]
```diff
@@ -235,6 +235,33 @@ Pros | Cons
...
+6. The `UI` displays the results in the result box.
+
+The following sequence diagram shows how the listing all student groups operation works:
+![Listing Student Groups Sequence Diagram](images/ListGroupSequenceDiagram.png)  
+                    
+#### 3.2.2 Design consideration:                                                
+                                                                                            
+##### Aspect: Retrieve all residents' student groups
+
+* **Alternative 1 (current choice):** Iterate through the resident list and then iterate through their student groups
```
Comment (jieffyy) [inline docs/DeveloperGuide.md]: Hmm... should we still keep the alternatives since there is only one design here?

File: docs/DeveloperGuide.md [hunk 3]
```diff
@@ -235,6 +235,33 @@ Pros | Cons
...
+### 3.2 Listing all student groups
+
+#### 3.2.1 Implementation
+The listing all student groups feature is facilitated by `ListGroupCommand`. It extends `Command` and overrides `Command#execute()` to list all student groups.
+
+Given below is a step-by-step usage scenario and how the listing all student groups feature works:
+
+1. The user launches the application and types `list-group` into the input box.
+2. The `UI` handles the input and calls `LogicManager#execute()` to execute it.
+3. The `AddressBookParser#parseCommand()` is called to parse the input and return a `ListGroupCommand`.
```
Comment (jieffyy) [inline docs/DeveloperGuide.md]: Should it be `......is called...and returns...`?

File: docs/DeveloperGuide.md (hunk 1 above)
Comment (teekoksiang) [reply docs/DeveloperGuide.md]: Yes, I should add that step. Thanks.

File: docs/DeveloperGuide.md (hunk 2 above)
Comment (teekoksiang) [reply docs/DeveloperGuide.md]: I was thinking about not writing it too. I will remove this then.

File: docs/DeveloperGuide.md (hunk 3 above)
Comment (teekoksiang) [reply docs/DeveloperGuide.md]: Noted, thanks.

File: docs/DeveloperGuide.md [hunk 4]
```diff
@@ -330,6 +327,24 @@ Pros | Cons
...
+#### 3.4.1 Implementation
+The listing all student groups feature is facilitated by `ListGroupCommand`. It extends `Command` and overrides `Command#execute()` to list all student groups.
+
+Given below is a step-by-step usage scenario and how the listing all student groups feature works:
+
+1. The user launches the application and types `list-group` into the input box.
+2. The `UI` handles the input and calls `LogicManager#execute()` to execute it.
+3. The `AddressBookParser#parseCommand()` is called to parse the input and returns a `ListGroupCommand`.
+4. The `ListGroupCommand` calls `ListGroupCommand#execute()` to retrieve the list of all residents by calling `Model#getFilteredPersonList()`.
+5. The `ListGroupCommand#execute()` iterates the list of all residents and then gets a set of all residents' student groups.
```
Comment (jieffyy) [inline docs/DeveloperGuide.md]: Sorry could you change this to make it clearer?

```markdown
4. The `ListGroupCommand` calls `ListGroupCommand#execute()` *which* retrieves the list of all residents by calling `Model#getFilteredPersonList()`.
5. The `ListGroupCommand#execute()` iterates *through* the list...
```

**Review by jieffyy — APPROVED** [review]
LGTM!

**Pull request merged**